fhir-bench-orchestrator/src/test_framework/mod.rs

```diff
@@ -212,14 +211,150 @@ impl std::fmt::Display for ServerOperationName {
+/// successfully.
+#[derive(Debug)]
+struct ServerOperationIterationSucceeded {
+    /// The state from the operation's completion.
+    completed: ServerOperationIterationCompleted,
+}
+
+/// This [ServerOperationIterationState] state node models an operation that failed to complete
+/// successfully.
+#[derive(Debug)]
+struct ServerOperationIterationFailed {
+    /// The state from the operation's completion.
+    completed: ServerOperationIterationCompleted,
+
+    /// The [anyhow::Error] detailing how/why the operation iteraion failed.
+    error: anyhow::Error,
+}
+
+impl ServerOperationIterationState<ServerOperationIterationStarting> {
+    /// Creates a new [ServerOperationIterationState] state machine instance, as the operation
+    /// iteration is being started.
+    pub fn new() -> ServerOperationIterationState<ServerOperationIterationStarting> {
+        ServerOperationIterationState {
+            _inner: ServerOperationIterationStarting {
+                started: Utc::now(),
+            },
+        }
+    }
+
+    /// Transitions this [ServerOperationIterationState] state machine instance after the operation
+    /// iteration completes, but before its success or failure has been determined.
+    pub fn completed(self) -> ServerOperationIterationState<ServerOperationIterationCompleted> {
+        ServerOperationIterationState {
+            _inner: ServerOperationIterationCompleted {
+                start: self._inner,
+                completed: Utc::now(),
+            },
+        }
+    }
+}
+
+impl ServerOperationIterationState<ServerOperationIterationCompleted> {
+    /// Transitions this [ServerOperationIterationState] state machine instance after the operation
+    /// iteration has been deemed a success.
+    pub fn succeeded(self) -> ServerOperationIterationState<ServerOperationIterationSucceeded> {
+        ServerOperationIterationState {
+            _inner: ServerOperationIterationSucceeded {
+                completed: self._inner,
+            },
+        }
+    }
+
+    /// Transitions this [ServerOperationIterationState] state machine instance after the operation
+    /// iteration has been deemed a failure.
+    ///
+    /// Parameters:
+    /// * `error`: the [anyhow::Error] detailing how/why the operation iteraion failed
+    pub fn failed(
+        self,
+        error: anyhow::Error,
+    ) -> ServerOperationIterationState<ServerOperationIterationFailed> {
+        ServerOperationIterationState {
+            _inner: ServerOperationIterationFailed {
+                completed: self._inner,
+                error,
+            },
+        }
+    }
+}
+
+impl ServerOperationIterationState<ServerOperationIterationSucceeded> {
+    /// Returns the [Duration] that the operation iteration ran for.
+    pub fn duration(&self) -> Duration {
+        self._inner.completed.completed - self._inner.completed.start.started
+    }
 }
```

Comment:
I really like all of this; it was fun to play with!

I'm definitely not 100% convinced it's worth all of the extra code, though. Something to ponder.

Regardless, leaving it in for now.